### PR TITLE
🌱 Add test for UpdateCoreDNS

### PR DESCRIPTION
**Pull request by wfernandes**
**What this PR does / why we need it**:
This PR backfills tests for UpdateCoreDNS for KCP.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Ref https://github.com/kubernetes-sigs/cluster-api/issues/3309

### CODE REVIEW

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 1]
```diff
@@ -29,11 +29,250 @@ import (
...
+					},
+				},
+			},
+			migrator: &fakeMigrator{
+				migrateErr: errors.New("failed to migrate"),
+			},
+			objs:      []runtime.Object{depl, cm, kubeadmCM},
+			expectErr: true,
+		},
+		// TODO: Add test to return error if updateCoreDNSDeployment fails.
```
Comment (wfernandes) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:255]: Not sure how to get helper.Patch to return error other than not having the deployment `depl` available. But we need the `depl` to exist for other cases above.

Comment (fabriziopandini) [reply controlplane/kubeadm/internal/workload_cluster_coredns_test.go:258]: Personally I take the following approach to solve this problem:
I first make unit tests for the atomic functions (in this case updateCoreDNSDeployment, updateCoreDNSCorefile etc) and exploit all the possible test cases while dealing with small portions of code/dependencies.
Then, when I test the higher level func (in this case UpdateCoreDNS) I'm only concerned about the small portion of code the is implemented in this func, because all the corner cases are already tested elsewhere.

In this case you can test updateCoreDNSDeployment, verify it manages properly images/volumes etc. and then come back to TestUpdateCoreDNS just test that updateCoreDNSDeployment is called

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go [hunk 2]
```diff
@@ -29,11 +29,253 @@ import (
...
```
Comment (fabriziopandini) [inline controlplane/kubeadm/internal/workload_cluster_coredns_test.go:258]: If I'm not wrong also the happy case is not tested (CoreDNSDeployment is upgraded)

File: controlplane/kubeadm/internal/workload_cluster_coredns_test.go (hunk 1 above)
Comment (wfernandes) [reply controlplane/kubeadm/internal/workload_cluster_coredns_test.go:258]: Yeah..I feel like that may be the best strategy for now. I see that there are tests for `updateCoreDNSDeployment` already so I'm confident with that.  
I'll remove this comment.

**Comment by fabriziopandini** [comment]
thanks for addressing the comments!
/lgtm
